File: python/circular-buffer/circular_buffer.py
class BufferFullException(Exception):
    pass


class BufferEmptyException(Exception):
    pass


class CircularBuffer(object):
    def __init__(self, size):
        self._size = size
        self._data = []

    def write(self, data):
        if len(self._data) >= self._size:
            raise BufferFullException('Buffer full')
        else:
            self._data.append(data)

    def read(self):
        if len(self._data):
            return self._data.pop(0)
        else:
            raise BufferEmptyException('Buffer empty')

    def overwrite(self, data):
        if len(self._data) < self._size:
            self.write(data)
        else:
            self._data = self._data[1::] + [data]

    def clear(self):
        self._data = []

File: javascript/circular-buffer/circular-buffer.js
function CircularBuffer(size) {
    let buffer = [];

    function read() {
        if (!buffer.length) {
            throw new BufferEmptyException();
        }
        return buffer.shift();
    }

    function write(data) {
        if (!data) {
            return;
        }
        if (buffer.length < size) {
            buffer.push(data);
        } else {
            throw new BufferFullException();
        }
    }

    function clear() {
        buffer = [];
    }

    function forceWrite(data) {
        if (buffer.length < size) {
            write(data);
        } else {
            buffer = buffer.slice(1)
            buffer.push(data);
        }

    }

    return {
        read: read,
        write: write,
        clear: clear,
        forceWrite: forceWrite
    };
}

function BufferEmptyException() {
    this.name = this.constructor.name;
    this.message = 'Buffer Empty';
}

function BufferFullException() {
    this.name = this.constructor.name;
    this.message = 'Buffer Full';
}

module.exports = {
    circularBuffer: function(capacity) {
        return new CircularBuffer(capacity);
    },

    bufferEmptyException: function() {
        return new BufferEmptyException();
    },

    bufferFullException: function() {
        return new BufferFullException();
    }
};